library/core/src/any.rs

```diff
@@ -1034,6 +1034,89 @@ impl<'a> Demand<'a> {
     pub fn is_satisfied(&self) -> bool {
         self.0.is_satisfied()
     }
+
+    /// Check if the `Demand` would be satisfied if provided with the
+    /// specified type. If the type does not match or has already been
+    /// provided, returns false.
+    ///
+    /// # Examples
+    ///
+    /// Check if an `u8` still needs to be provided and then provides
+    /// it.
+    ///
+    /// ```rust
+    /// # #![feature(provide_any)]
+    /// use std::any::{Provider, Demand};
+    /// # struct SomeConcreteType;
+    /// # fn expensive_computation() -> Option<u8> { Some(42) }
+    ///
+    /// impl Provider for SomeConcreteType {
+    ///     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    ///         if demand.would_be_satisfied_by_value_of::<u8>() {
+    ///             if let Some(value) = expensive_computation() {
+    ///                 demand.provide_value(value);
+    ///             }
+    ///         }
+    ///         if demand.would_be_satisfied_by_value_of::<u8>() {
+    ///             unreachable!("An u8 has already been provided");
+    ///         }
+    ///     }
+    /// }
+    /// # let sct = SomeConcreteType;
+    /// # std::any::request_value::<u8>(&sct);
```

Comment:
should this be an assert?

Reply:
I don't see why it couldn't be, but the "real assertion" is that the `unreachable` isn't called. It also feels a little odd to have an assertion that's doc-hidden.

I'd be more than happy to reorganize these to be more obvious, but my brain couldn't come up with something.

Reply:
I think my default would be to not doc hide this, have it be the "real assertion", show the definition of the type and make it's implementation depend on the instance you construct with it, construct an instance that will definitely return a specific value and then request that value and assert its the one you expect.

I'm not sure I explained this very clearly so I can hack together a quick example if you need

Reply:
Took a crack at what I think you meant; let me know how close I got!